test/functional/interface_bitcoin_cli.py

```diff
@@ -107,6 +108,33 @@ def run_test(self):
         self.log.info("Test connecting to a non-existing server")
         assert_raises_process_error(1, "Could not connect to the server", self.nodes[0].cli('-rpcport=1').echo)
 
+        self.log.info("Test handling of invalid ports in rpcconnect")
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:notaport", self.nodes[0].cli("-rpcconnect=127.0.0.1:notaport").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:-1", self.nodes[0].cli("-rpcconnect=127.0.0.1:-1").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:0", self.nodes[0].cli("-rpcconnect=127.0.0.1:0").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:65536", self.nodes[0].cli("-rpcconnect=127.0.0.1:65536").echo)
```

Comment:
Thanks @kevkevinpal.  Good example of why review is important.

Your findings nudged me to add some assert test statements to help cover cases involving IPv6 hosts.

We're experiencing the result of the scope of `SplitHostPort()`.  `SplitHostPort()` extracts the host and port.  Although the port is validated, host is not (and the function declaration comment header mentions validating the port, but doesn't mention validating the host).

https://github.com/bitcoin/bitcoin/blob/7fee0ca014b1513e836d2550d1b7512739d5a79a/src/util/strencodings.h#L97-L106

`SplitHostPort()` is treating "127.0.0.1::" as an IPv6 host since "::" is used for IPv6 address shortening.  `SplitHostPort()` detects no port as part of the input string, which would be why `Could not connect to the server 127.0.0.1:::18443` is the error provided for `src/bitcoin-cli -rpcconnect=127.0.0.1:: getblockcount` (the ":18443" is tacked on the end because it's the default port for regtest).

An enhancement would be to perform a validation of the host received from `SplitHostPort()` before bitcoin-cli attempts to contact the host.  I'm thinking we would need to first define what is "valid" and what is "invalid" for use with bitcoin-cli.  The gist of this kind of checking seems to be present in `CNetAddr`.  In general, I would think bitcoin-cli shouldn't be overly restrictive in what it lets the user attempt to use as a host.

https://github.com/bitcoin/bitcoin/blob/7fee0ca014b1513e836d2550d1b7512739d5a79a/src/netaddress.h#L157-L181

For now, it might make the most sense to leave the enhancement of host validation to a new PR since this PR is focused on port error detection.